topics/client-custom-plugins.md

```diff
@@ -0,0 +1,269 @@
+
+The [previous section](#first-plugin) demonstrates how to create a plugin that appends a predefined custom header to each response. Let's make this plugin more useful and provide a configuration for passing any custom header name and value:
+
+1. First, you need to define a configuration class:
+
+   ```kotlin
+   ```
+   {src="snippets/client-custom-plugin/src/main/kotlin/com/example/plugins/CustomHeaderConfigurable.kt" include-lines="14-17"}
+
+2. To use this configuration in a plugin, pass a configuration class reference to `createApplicationPlugin`:
+
+   ```kotlin
+   ```
+   {src="snippets/client-custom-plugin/src/main/kotlin/com/example/plugins/CustomHeaderConfigurable.kt" include-lines="3-12"}
+
+   Given that plugin configuration fields are mutable, saving them in local variables is recommended.
+
+3. Finally, you can install and configure the plugin as follows:
+
+   ```kotlin
+   ```
+   {src="snippets/client-custom-plugin/src/main/kotlin/com/example/Application.kt" include-lines="11-15,18"}
+
+> You can find the full example here: [CustomHeaderConfigurable.kt](https://github.com/ktorio/ktor-documentation/blob/%ktor_version%/codeSnippets/snippets/client-custom-plugin/src/main/kotlin/com/example/plugins/CustomHeaderConfigurable.kt).
+
+
+
+## Handle requests and responses {id="call-handling"}
+
+Custom plugins provide access to different stages of handling requests and responses 
+using a set of dedicated handlers, for example:
+- `onRequest` and `onResponse` allow you to handle requests and responses, respectively.
+- `transformRequestBody` and `transformResponseBody` can be used to apply necessary transformations to 
+   request and response bodies.
+
+There is also the `on(...)` handler that allows you to invoke specific hooks that might be useful to handle other stages of a call.
+The table below lists all handlers in the order they are executed:
+
+<table>
+<tr>
+<td>
+Handler
+</td>
+<td>
+Description
+</td>
+</tr>
+
+<tr>
+<td>
+<code>on(SetupRequest)</code>
+</td>
+<td>
+The <code>SetupRequest</code> hook is executed first in request processing.
+</td>
+</tr>
+
+<tr>
+<td>
+<code>onRequest</code>
+</td>
+<td>
+This handler is executed for each HTTP <a href="request.md">request</a> and allows you to modify it.
+The <a anchor="example-custom-header"/> example shows how to append a custom header to each request by handling <code>onRequest</code>.
+</td>
+</tr>
+
+<tr>
+<td>
+<code>transformRequestBody</code>
+</td>
+<td>
+Allows you to transform a <a href="request.md" anchor="body">request body</a>.
+In this handler, you need to serialize the body into 
+<a href="https://api.ktor.io/ktor-http/io.ktor.http.content/-outgoing-content/index.html">OutgoingContent</a> 
+(for example, <code>TextContent</code>, <code>ByteArrayContent</code>, or <code>FormDataContent</code>)
+or return <code>null</code> if your transformation is not applicable.
+</td>
+</tr>
+
+
+<tr>
+<td>
+<code>on(Send)</code>
```

Comment:
What do you think about separating basic callbacks from hooks? Up to you, but I think it may be easier for people to understand first basic API and then learn about hooks.

Reply:
Good point, thanks. I thought about it but still didn't find a good solution to structure the topic in this case. I'll file a corresponding issue and will make another try a bit later.